### PR TITLE
Add a foreign layer URL host whitelist

### PR DESCRIPTION
**Pull request by nwt**
Until we have some experience hosting foreign layer manifests, the Hub
operators wish to limit foreign layers on Hub to those hosted by
Microsoft. To that end, this change adds a registry configuration option
to whitelist hosts that may appear in foreign layer URLs of pushed
manifests, and it modifies manifest validation to reject pushes
containing foreign layer URLs whose host is not whitelisted.

Signed-off-by: Noah Treuhaft <noah.treuhaft@docker.com>

### CODE REVIEW

**Comment by dmp42** [comment]
cc @runcom @docker/core-distribution-maintainers 

and related to #1825

**Comment by runcom** [comment]
seems inline with #1825 so :+1:

File: docs/configuration.md [hunk 1]
```diff
@@ -491,6 +495,27 @@ map.
...
+match a list entry. Matching is case-insensitive, and domains entries
+match any subdomain while host entries must match exactly.
+
+The following example allows `example.com` and `WWW.EXAMPLE.NET` but not
+`WWW.EXAMPLE.COM` or `www.example.net`.
+
+    foreignlayer:
+        hostwhitelist: [EXAMPLE.COM, .example.net]
+
+Set `hostwhitelist` to an empty list (`[]`) to reject all foreign-layer
```
Comment (RichardScothern) [inline docs/configuration.md]: could you just add an `enabled: <bool>` under the `foreignlayer` section instead?  (more explicit and easier for testing)

File: docs/configuration.md [hunk 2]
```diff
@@ -401,6 +403,8 @@ Permitted values are `error`, `warn`, `info` and `debug`. The default is
         enabled: false
       cache:
         blobdescriptor: inmemory
+      foreignlayer:
+        hostwhitelist: [example.com, example.net]
```
Comment (RichardScothern) [inline docs/configuration.md]: Maybe just `whitelist` since there are domains?

File: registry/handlers/app.go [hunk 3]
```diff
@@ -194,6 +195,28 @@ func NewApp(ctx context.Context, config *configuration.Configuration) *App {
 		}
 	}
 
+	// configure foreign layers
+	if flConfig, ok := config.Storage["foreignlayers"]; ok {
+		if hosts, ok := flConfig["hostwhitelist"]; ok {
+			switch hosts := hosts.(type) {
+			case []interface{}:
```
Comment (RichardScothern) [inline registry/handlers/app.go]: Is this usable with environment variables? e.g.
REGISTRY_STORAGE_FOREIGNLAYER_HOSTWHITELIST="[a.com]"

File: docs/configuration.md [hunk 4]
```diff
@@ -491,6 +495,27 @@ map.
...
+host is allowed.)
+
+Set the `hostwhitelist` field to a list of domains (strings beginning
+with `.`) and hosts (strings not beginning with `.`). Pushing a manifest
+will then succeed only if all foreign layer URLs contain hosts that
+match a list entry. Matching is case-insensitive, and domains entries
+match any subdomain while host entries must match exactly.
+
+The following example allows `example.com` and `WWW.EXAMPLE.NET` but not
+`WWW.EXAMPLE.COM` or `www.example.net`.
```
Comment (RichardScothern) [inline docs/configuration.md]: example.net should be accepted.

File: registry/handlers/app.go [hunk 5]
```diff
@@ -194,6 +195,28 @@ func NewApp(ctx context.Context, config *configuration.Configuration) *App {
...
+				hostStrings := make([]string, len(hosts))
+				for i, host := range hosts {
+					switch host := host.(type) {
+					case string:
+						hostStrings[i] = host
```
Comment (RichardScothern) [inline registry/handlers/app.go]: Check for valid host/domain format?

File: registry/handlers/app.go (hunk 3 above)
Comment (aaronlehmann) [inline registry/handlers/app.go]: With the current code I don't think this would work, but you could probably work around it with something like `REGISTRY_STORAGE_FOREIGNLAYER="hostwhitelist: [a.com]"`.

We don't have a good general solution for expressing array values in environment variables.

File: docs/configuration.md (hunk 1 above)
Comment (nwt) [inline docs/configuration.md]: Certainly. And in that case, panicking on an empty whitelist seems appropriate.

File: docs/configuration.md (hunk 2 above)
Comment (nwt) [inline docs/configuration.md]: Will do.

File: registry/handlers/app.go (hunk 3 above)
Comment (nwt) [inline registry/handlers/app.go]: Is this good as is or should `REGISTRY_STORAGE_FOREIGNLAYER_HOSTWHITELIST='[a.com]` be made to work?

File: docs/configuration.md (hunk 4 above)
Comment (nwt) [inline docs/configuration.md]: Agreed.

File: registry/handlers/app.go (hunk 3 above)
Comment (aaronlehmann) [inline registry/handlers/app.go]: IMHO supporting that variation is out of scope. We should test to ensure something like `REGISTRY_STORAGE_FOREIGNLAYER="hostwhitelist: [a.com]"` is possible, though.

File: registry/handlers/app.go (hunk 5 above)
Comment (nwt) [inline registry/handlers/app.go]: Happy to add that if it's wanted.

File: registry/handlers/app.go (hunk 3 above)
Comment (RichardScothern) [inline registry/handlers/app.go]: Yeah, I'm not sure if what you will get will be a `[]interface`, if it comes from an environment variable is may be `[]string`

Comment (aaronlehmann) [inline registry/handlers/app.go]: Environment variables are always fed through the YAML parser, so I don't think this is an issue. Again, it's worth testing this.

File: configuration/configuration.go [hunk 6]
```diff
@@ -350,6 +350,8 @@ func (storage Storage) Type() string {
 			// allow configuration of caching
 		case "delete":
 			// allow configuration of delete
+		case "foreignlayers":
```
Comment (stevvooe) [inline configuration/configuration.go]: What does this have to do with storage? This should be a registry option.

File: docs/configuration.md (hunk 1 above)
Comment (stevvooe) [inline docs/configuration.md]: Let's use the term `allowed`.

Comment (stevvooe) [inline docs/configuration.md]: Or better yet:

```
foreignlayer:
  hosts:
    allow:
      - www.foo.com
      - example.com
```

File: docs/configuration.md [hunk 7]
```diff
@@ -491,6 +495,27 @@ map.
 >are equivalent, `layerinfo` has been deprecated, in favor or
 >`blobdescriptor`.
 
+### foreignlayers
```
Comment (stevvooe) [inline docs/configuration.md]: This shouldn't really be called "Foreign Layers". The registry doesn't and shouldn't know what a foreign layer is.

This feature influences the validation of manifest urls.

File: docs/configuration.md (hunk 4 above)
Comment (stevvooe) [inline docs/configuration.md]: I don't understand these limitations. Do we need to employ https://godoc.org/golang.org/x/net/publicsuffix? What are you trying to do?

**Comment by stevvooe** [comment]
This feature shouldn't be called "Foreign Layers". The concept of a foreign layer is reserved to the docker engine, as the layer is foreign to docker. The registry doesn't care. This just influences the valid urls that a registry will accept in the manifest.

File: configuration/configuration.go (hunk 6 above)
Comment (stevvooe) [inline configuration/configuration.go]: I see why this is a logical place. In the past, we had `Compatibility` to include manifest validation, which was top-level.

**Comment by runcom** [comment]
Can we take care of enabling/disabling the whole foreign layer story/feature as part of this PR? Related to poing 3) in https://github.com/docker/distribution/issues/1825

Is it feasible in this PR? seems likely to me, just an `enabled` flag under `foreignLayers:` opt (or whatever it'll be called)
If not, I'll take care of it as a follow up (but I don't see issues in doing this here.

The point from https://github.com/docker/distribution/issues/1825 is:
> 3) provide a way to opt-in for such a feature at the registry level
Since there may be the need to defer this support on some deployment, we can have an explicit setting which opt-in for this new behavior.

@dmp42 @stevvooe wdyt?

File: registry/storage/registry.go [hunk 8]
```diff
@@ -154,6 +171,20 @@ func (reg *registry) BlobStatter() distribution.BlobStatter {
 	return reg.statter
 }
 
+// ForeignLayerHostWhitelist tests whether the host is allowed in foreign layer URLs.
+func (reg *registry) ForeignLayerHostAllowed(host string) bool {
+	if reg.foreignLayerHostWhitelist == nil {
```
Comment (runcom) [inline registry/storage/registry.go]: the knob I'm talking about can cleanly take care of this check also - instead of checking for the presence of a whitelist, we can check if foreign layers are enabled at all and exit early here

File: docs/configuration.md (hunk 4 above)
Comment (nwt) [inline docs/configuration.md]: I'm trying to separate the whitelist-a-single-name operation from the whitelist-all-subdomains-of-a-name operation.

**Comment by nwt** [comment]
> This feature shouldn't be called "Foreign Layers". 

I'm happy follow maintainer guidance on an appropriate name. I don't think I have enough perspective to propose an alternative. Except for "Not Foreign Layers". 💥

File: docs/configuration.md (hunk 1 above)
Comment (nwt) [inline docs/configuration.md]: Happy to follow either this proposal or @RichardScothern's proposal  of `whitelist`.

**Comment by nwt** [comment]
> Is it feasible in this PR? seems likely to me, just an `enabled` flag under `foreignLayers:` opt (or whatever it'll be called)

@RichardScothern proposed [the same](https://github.com/docker/distribution/pull/1829#discussion_r70340968) and I'm planning to implement it.

**Comment by runcom** [comment]
> Is it feasible in this PR? seems likely to me, just an enabled flag under foreignLayers: opt (or whatever it'll be called)
> @RichardScothern proposed the same and I'm planning to implement it.

ty

File: docs/configuration.md (hunk 4 above)
Comment (stevvooe) [inline docs/configuration.md]: I see. Leading `.` is the wildcard. Are there other systems that follow this convention?

**Comment by stevvooe** [comment]
@nwt This is validating manifest urls:

```
validation:
  disabled: false
  manifests:
    urls:
      allowed:
        - https://foo.com
```

A few things:

1. This is configuring how we validate incoming manifests. While this is on storage configuration, let's consider pulling the validation above storage.
2. We should also think about prevention of serving manifests with disallowed urls. For example, if it doesn't pass url validation, should we serve it at all? This is important since we don't block these _at all_ now.
3. We should be validating URLs, not hosts and domains.

File: docs/configuration.md (hunk 4 above)
Comment (nwt) [inline docs/configuration.md]: http://httpd.apache.org/docs/current/mod/mod_proxy.html#noproxy

**Comment by nwt** [comment]
@stevvooe: I implemented this:
```
validation:
  manifests:
    urls:
      disabled: false
      allowed: regexp
```
I changed `allowed` to a regular expression because Hub needs more control than a list of strings can provide, and I moved `disabled` under `urls` since that's the only validation it influences.

**Comment by runcom** [comment]
I meant to have disabled (or enabled) as a way to fully opt-out this features. It seems what you did doesn't cleanly achieve that, does it?

**Comment by nwt** [comment]
@runcom: I think it does. When

**Comment by nwt** [comment]
@runcom: All pushes of foreign layer manifests will be rejected with this in your config:
```
validation:
  manifests:
    urls:
      disabled: true
```

Were you asking for something different?

**Comment by runcom** [comment]
That's somehow fine ti me, i was hoping for such a flag to be more global but if it still accomplishes fully opt-out than I'll leave those decisions to others about where that disable flag should be.

File: configuration/configuration.go [hunk 9]
```diff
@@ -172,6 +172,15 @@ type Configuration struct {
 			TrustKey string `yaml:"signingkeyfile,omitempty"`
 		} `yaml:"schema1,omitempty"`
 	} `yaml:"compatibility,omitempty"`
+
+	Validation struct {
+		Manifests struct {
+			URLs struct {
+				Allowed  *string `yaml:"allowed,omitempty"`
```
Comment (stevvooe) [inline configuration/configuration.go]: Why not a list?

File: configuration/configuration.go [hunk 10]
```diff
@@ -172,6 +172,15 @@ type Configuration struct {
 			TrustKey string `yaml:"signingkeyfile,omitempty"`
 		} `yaml:"schema1,omitempty"`
 	} `yaml:"compatibility,omitempty"`
+
+	Validation struct {
```
Comment (stevvooe) [inline configuration/configuration.go]: Not sure how this passes the linter, but please document the field.

File: docs/configuration.md [hunk 11]
```diff
@@ -1796,6 +1801,51 @@ defines a such a feature with configurable behavior.
...
+      manifests:
+        urls:
+          allowed: ^https?://(www\.)?example\.com
+          disabled: false
+
+### Manifests
+
+Use the `manifest` subsection to configure manifest validation.
+
+#### URLs
```
Comment (stevvooe) [inline docs/configuration.md]: Make this clear that this applies to any descriptors that have urls in a manifest. This validation needs to include any descriptor in manifest list, as well.

File: configuration/configuration.go (hunk 10 above)
Comment (aaronlehmann) [inline configuration/configuration.go]: I believe the linter doesn't care about documentation for nested structs, only structs defined at the top level.

File: registry/storage/registry.go [hunk 12]
```diff
@@ -46,6 +52,23 @@ func DisableDigestResumption(registry *registry) error {
 	return nil
 }
 
+// DisableManifestURLs is a functional option for NewRegistry. It causes the
+// registry to reject pushed manifests containing URLs.
+func DisableManifestURLs(registry *registry) error {
```
Comment (stevvooe) [inline registry/storage/registry.go]: O, these are separate features? Why not just have a `deny` section?

```
validation:
  enabled: true # enabled turns on the feature
  manifests:
    urls:
      deny:
        - *
```

Typically, we use `enabled`/`disabled` as a feature switch. Having mean that "urls are disabled" is confusing.

**Comment by runcom** [comment]
@stevvooe +1

**Comment by nwt** [comment]
@stevvooe: 153c2f2cb946839a10837d771111658a89e9234d should be closer to what you had in mind than 2e376295a135821963afd7c6eae2c200b30dbbb6 was, with `enabled` immediately below `validation`, `deny` as well as `allow`, and lists for both.

File: docs/configuration.md [hunk 13]
```diff
@@ -1796,6 +1802,39 @@ defines a such a feature with configurable behavior.
   </tr>
 </table>
 
+## Validation
+
+    validation:
+      disabled: false
+      manifests:
+        urls:
+          allow: ^https?://([^/]+\.)*example\.com/
```
Comment (stevvooe) [inline docs/configuration.md]: I think these examples need to be a list.

File: configuration/configuration.go [hunk 14]
```diff
@@ -172,6 +172,24 @@ type Configuration struct {
 			TrustKey string `yaml:"signingkeyfile,omitempty"`
 		} `yaml:"schema1,omitempty"`
 	} `yaml:"compatibility,omitempty"`
+
+	// Validation configures validation options for the registry.
+	Validation struct {
+		// Disabled disables all other options in this section.
+		Disabled bool `yaml:"disabled,omitempty"`
```
Comment (stevvooe) [inline configuration/configuration.go]: By default, this should be "disabled". The zero-value of `Disabled`, would be `false`, which would mean `Enabled`. This field should be called `Enabled` and users should be required to set it to true to enable the validation.

I think, by default, we can refuse manifests with anything in the url field.

**Comment by stevvooe** [comment]
LGTM

There are a few nits about getting the defaults backwards compatible.

Great addition, @nwt!

**Comment by nwt** [comment]
@stevvooe: 61e5803b56a41292779bdd259ea14f166531579e addresses comments on 153c2f2cb946839a10837d771111658a89e9234d.

1. Documentation examples properly show `allow` and `deny` as lists.
1. `Configuration.Validation.Enabled` replaces `….Disabled`.
1. If `allow` and `deny` are empty, all manifests with URLs are rejected.


I'm not thrilled with how I handled for that last one, both because`deny` is no longer useful without `allow` and because `storage.ManifestURLsAllowRegexp(regexp.MustCompile("^$"))` in app.go is a bit magical, so let me know if you think either of those edges needs filing.

**Comment by stevvooe** [comment]
@nwt 

> I'm not thrilled with how I handled for that last one, both becausedeny is no longer useful without allow and because `storage.ManifestURLsAllowRegexp(regexp.MustCompile("^$"))` in app.go is a bit magical, so let me know if you think either of those edges needs filing.

Will it still allow manifests _without_ urls? That is somewhat the purpose. For example, if one configures the following:

```
validation:
  enabled: true
```

All manifests with urls would be blocked, while others would be allowed.

Let's say we unrolled these into an ordered set of operations. By default, with all empy `allow` and `deny`, it would look like this:

```
DENY ALL
```

This would give us default closed operation when the URL is empty.

We insert all allows before denies:

```
ALLOW foo.com
DENY ALL
```

ALLOW always has precedent:

```
ALLOW foo.com
DENY foo.com
DENY ALL
```

These are matched such that if the rule matches, we return the operation. For `foo.com`, we'd match an `ALLOW` rule first and the URL would be allowed.

This only becomes problematic when we have a more specific deny than an allow:

```
ALLOW .*
DENY foo.com
DENY ALL
```

So, even though we deny `foo.com`, allow has precedent and is match before the deny. We can fix this by ordering DENY before allow, but we get the same issue, except we have allow that is more specific than a deny:

```
DENY .*
ALLOW foo.com
DENY ALL
```

Typically, this issue is dealt with using priorities, but that is overbuilt for this purpose. It is probably better to favor ALLOW for this purpose but arguments can go either way.

__TL; DR__ Having `^$` as a "magic" deny entry is an okay bit of magic. That defines the default operation on the match chain.

**Comment by nwt** [comment]
@stevvooe:

> Will it still allow manifests without urls?

Yes, manifests without URLs are always allowed regardless of what appears under `validation`.

**Comment by RichardScothern** [comment]
LGTM

**Comment by stevvooe** [comment]
LGTM

**Comment by dmcgowan** [comment]
LGTM

**Comment by RichardScothern** [comment]
Holding off on merging while we discuss a missing API test..

**Comment by RichardScothern** [comment]
Fixes a pre-existing condition and coming in a PR.
